From 7247fa1a4e8e83888eaf56529c0f96350fceaf34 Mon Sep 17 00:00:00 2001
From: Sammy He <r62914@freescale.com>
Date: Wed, 21 Dec 2011 20:03:12 +0800
Subject: [PATCH 188/280] ENGR00170748 vpu: Fix MJPG decoder with tiled format
 failure issue

Fix MJPG decoder with tiled format failure issue, and add the error
handler when JPG header data is corrupted.

Signed-off-by: Sammy He <r62914@freescale.com>
---
 vpu/vpu_lib.c  |  4 ++++
 vpu/vpu_util.c | 12 ++++++++----
 2 files changed, 12 insertions(+), 4 deletions(-)

diff --git a/vpu/vpu_lib.c b/vpu/vpu_lib.c
index b8cd0d3..dcee81e 100644
--- a/vpu/vpu_lib.c
+++ b/vpu/vpu_lib.c
@@ -2729,9 +2729,13 @@ RetCode vpu_DecGetInitialInfo(DecHandle handle, DecInitialInfo * info)
 	pDecInfo = &pCodecInst->CodecInfo.decInfo;
 
 	if (is_mx6q_mjpg_codec(pCodecInst->codecMode)) {
+
 		if (!LockVpu(vpu_semap))
 			return RETCODE_FAILURE_TIMEOUT;
 
+		/* set tiled map type info */
+		SetTiledMapTypeInfo(pDecInfo->mapType, &pDecInfo->sTiledInfo);
+
 		if (pDecInfo->jpgInfo.lineBufferMode) {
 			pDecInfo->jpgInfo.pVirtJpgChunkBase = pDecInfo->jpgInfo.pVirtBitStream;
 			pDecInfo->jpgInfo.chunkSize = pDecInfo->streamBufSize;
diff --git a/vpu/vpu_util.c b/vpu/vpu_util.c
index 66c746d..28205d3 100644
--- a/vpu/vpu_util.c
+++ b/vpu/vpu_util.c
@@ -2007,13 +2007,15 @@ int decode_sof_header(JpgDecInfo *jpg)
 	}
 
 	numComp = get_bits(&jpg->gbc, 8);
-	if (numComp > 3)
-		info_msg("Picture Horizontal Size limits Maximum size\n");
+	if (numComp > 3) {
+		info_msg("Components number limits Maximum size: numComp %d\n", numComp);
+		return 0;
+	}
 
 	if (get_bits_left(&jpg->gbc) < (numComp * ( 8 + 4 + 4 + 8) + 24))
 		return 0;
 
-	for (i=0; i<numComp; i++) {
+	for (i = 0; i < numComp; i++) {
 		compID = get_bits(&jpg->gbc, 8);
 		hSampFact[i] = get_bits(&jpg->gbc, 4);
 		vSampFact[i] = get_bits(&jpg->gbc, 4);
@@ -2027,8 +2029,10 @@ int decode_sof_header(JpgDecInfo *jpg)
 
 	if ((hSampFact[0] > 2) || (vSampFact[0] > 2) || ((numComp == 3) &&
 	    ((hSampFact[1] != 1) || (hSampFact[2] != 1) ||
-	     (vSampFact[1] != 1) || (vSampFact[2] != 1))))
+	     (vSampFact[1] != 1) || (vSampFact[2] != 1)))) {
 		info_msg("Not Supported Sampling Factor\n");
+		return 0;
+	}
 
 	if (numComp == 1)
 		sampleFactor = SAMPLE_400;
-- 
1.8.0

